libs/vkd3d/swapchain.c

```diff
@@ -0,0 +1,2216 @@
+                &output)); ++output_idx)
+        {
+            if (FAILED(hr = IDXGIOutput_GetDesc(output, &desc)))
+            {
+                WARN("Adapter %u output %u: Failed to get output desc, hr %#x.\n", adapter_idx,
+                        output_idx, hr);
+                IDXGIOutput_Release(output);
+                continue;
+            }
+
+            if (desc.Monitor == monitor)
+            {
+                *dxgi_output = output;
+                IDXGIAdapter_Release(adapter);
+                return S_OK;
+            }
+
+            IDXGIOutput_Release(output);
+        }
+        IDXGIAdapter_Release(adapter);
+    }
+
+    if (hr != DXGI_ERROR_NOT_FOUND)
+        WARN("Failed to enumerate outputs, hr %#x.\n", hr);
+
+    WARN("Output could not be found.\n");
+    return DXGI_ERROR_NOT_FOUND;
+}
+
+struct d3d12_swapchain
+{
+    IDXGISwapChain3 IDXGISwapChain3_iface;
+    LONG refcount;
+    struct vkd3d_private_store private_store;
+
+    VkSwapchainKHR vk_swapchain;
+    VkSurfaceKHR vk_surface;
+    VkFence vk_fence;
+    VkDeviceMemory vk_memory[DXGI_MAX_SWAP_CHAIN_BUFFERS];
+    VkCommandPool vk_cmd_pool;
+    VkImage vk_images[DXGI_MAX_SWAP_CHAIN_BUFFERS];
+    VkImage vk_swapchain_images[DXGI_MAX_SWAP_CHAIN_BUFFERS];
+    VkCommandBuffer vk_cmd_buffers[DXGI_MAX_SWAP_CHAIN_BUFFERS];
+    VkSemaphore vk_semaphores[DXGI_MAX_SWAP_CHAIN_BUFFERS];
+    ID3D12Resource *buffers[DXGI_MAX_SWAP_CHAIN_BUFFERS];
+    unsigned int buffer_count;
+    unsigned int vk_swapchain_width;
+    unsigned int vk_swapchain_height;
+    VkPresentModeKHR present_mode;
+
+    uint32_t vk_image_index;
+    unsigned int current_buffer_index;
+
+    struct d3d12_command_queue *command_queue;
+    IDXGIFactory *factory;
+
+    HWND window;
+    IDXGIOutput *target;
+    DXGI_SWAP_CHAIN_DESC1 desc;
+    DXGI_SWAP_CHAIN_FULLSCREEN_DESC fullscreen_desc;
+
+    ID3D12Fence *frame_latency_fence;
+    HANDLE frame_latency_event;
+
+    uint64_t frame_number;
+    uint32_t frame_latency;
+};
+
+static inline const struct vkd3d_vk_device_procs* d3d12_swapchain_procs(struct d3d12_swapchain* swapchain)
+{
+    return &swapchain->command_queue->device->vk_procs;
+}
+
+static inline struct ID3D12Device* d3d12_swapchain_device_iface(struct d3d12_swapchain* swapchain)
+{
+    return &swapchain->command_queue->device->ID3D12Device_iface;
+}
+
+static inline struct d3d12_device* d3d12_swapchain_device(struct d3d12_swapchain* swapchain)
+{
+    return swapchain->command_queue->device;
+}
+
+static inline struct ID3D12CommandQueue* d3d12_swapchain_queue_iface(struct d3d12_swapchain* swapchain)
+{
+    return &swapchain->command_queue->ID3D12CommandQueue_iface;
+}
+
+static HRESULT d3d12_swapchain_set_display_mode(struct d3d12_swapchain *swapchain,
+        IDXGIOutput* output, DXGI_MODE_DESC *mode)
+{
+    FIXME("TODO");
+    return S_OK;
+}
+
+static HRESULT d3d12_swapchain_set_fullscreen(struct d3d12_swapchain *swapchain)
+{
+    FIXME("TODO!");
```

Comment:
Does this work in Wine?

Reply:
Partially... but I have now done a proper and full implementation of fullscreen.